src/casts/specify-representation.rkt

```diff
@@ -855,32 +834,30 @@ but a static single assignment is implicitly maintained.
                                (cons "argument" e))
                              e*))))]
         [(Make-Fn-Coercion mk-crcn (app recur t1) (app recur t2) (app recur l))
-         (: invoke-mk-fn-crcn ((Code-Label Uid) (Var Uid) D0-Expr D0-Expr -> D0-Expr))
-         (define (invoke-mk-fn-crcn mk-fn t1 t2 l)
-           (App-Code
-            mk-fn
-            (list t1 t2 l ZERO-IMDT
-                  (sr-tagged-array-ref t1 TYPE-FN-TAG TYPE-FN-ARITY-INDEX))))
-         (let ([mk-fn-crcn (get-mk-fn-crcn! mk-crcn)])
-           (if (Var? t1)
-               (invoke-mk-fn-crcn mk-fn-crcn t1 t2 l)
-               (begin$
-                 (assign$ fn-type1 t1)
-                 (invoke-mk-fn-crcn mk-fn-crcn fn-type1 t2 l))))]
-        [(Compose-Fn-Coercion compose (app recur c1) (app recur c2))
-         (: invoke-comp ((Code-Label Uid) (Var Uid) D0-Expr -> D0-Expr))
-         (define (invoke-comp comp-fn c1 c2)
-           (begin$
-             (assign$ tagged-arity
-               (sr-tagged-array-ref
-                c1 COERCION-MEDIATING-TAG COERCION-FN-ARITY-INDEX))
-             (assign$ arity (op$ %>> tagged-arity COERCION-SECOND-TAG-SHIFT))
-             (App-Code comp-fn (list c1 c2 ZERO-IMDT arity TRUE-IMDT))))
-         (let ([mk-fn-crcn (get-comp-fn-crcn! compose)])
-           (if (Var? c1)
-               (invoke-comp mk-fn-crcn c1 c2)
-               (begin$
-                 (assign$ fn-crcn1 c1) (invoke-comp mk-fn-crcn fn-crcn1 c2))))]
+         (define mk-fn-crcn (get-mk-fn-crcn! mk-crcn))
+         (: invoke-mk-fn-crcn ((Var Uid) (Var Uid) -> D0-Expr))
+         (define (invoke-mk-fn-crcn t1 t2) (app-code$ mk-fn-crcn t1 t2 l))
+         (cond
+           [(and (Var? t1) (Var? t2)) (invoke-mk-fn-crcn t1 t2)]
+           [(Var? t1) (begin$ (assign$ a-t2 t2) (invoke-mk-fn-crcn t1 a-t2))]
+           [(Var? t2) (begin$ (assign$ a-t1 t1) (invoke-mk-fn-crcn a-t1 t2))]
+           [else (begin$
+                   (assign$ a-t1 t1)
+                   (assign$ a-t2 t2)
+                   (invoke-mk-fn-crcn a-t1 a-t2))])]
+        [(Compose-Fn-Coercion comp-crcn (app recur c1) (app recur c2))
+         (define comp-fn-crcn (get-comp-fn-crcn! comp-crcn))
+         (: invoke-comp-fn-crcn ((Var Uid) (Var Uid) -> D0-Expr))
+         (define (invoke-comp-fn-crcn c1 c2)
+           (app-code$ comp-fn-crcn c1 c2))
+         (cond
+           [(and (Var? c1) (Var? c2)) (invoke-comp-fn-crcn c1 c2)]
+           [(Var? c1) (begin$ (assign$ a-c2 c2) (invoke-comp-fn-crcn c1 a-c2))]
+           [(Var? c2) (begin$ (assign$ a-c1 c1) (invoke-comp-fn-crcn a-c1 c2))]
+           [else (begin$
+                   (assign$ a-c1 c1)
+                   (assign$ a-c2 c2)
+                   (invoke-comp-fn-crcn a-c1 a-c2))])]
```

Comment:
This chunk of code, that is doing manual ANF, doesn't seem to be necessary.

Reply:
Done.